src/expr/core/src/expr/expr_some_all.rs

```diff
@@ -160,12 +160,17 @@ impl Expression for SomeAllExpression {
         );
 
         let func_results = self.func.eval(&data_chunk).await?;
-        let mut func_results_iter = func_results.as_bool().iter();
+        let bools = func_results.as_bool();
+        let mut offset = 0;
         Ok(Arc::new(
             num_array
                 .into_iter()
                 .map(|num| match num {
-                    Some(num) => self.resolve_bools(func_results_iter.by_ref().take(num)),
+                    Some(num) => {
+                        let range = offset..offset + num;
+                        offset += num;
+                        self.resolve_bools(range.map(|i| bools.value_at(i)))
+                    }
```

Comment:
So the actual bug is that we mistakenly expect `resolve_bools` to exhaust the passed-in iterator, right? This is quite subtle. 🤣 Could you please add some comments to elaborate this?

Reply:
Yes. I didn't realize the passed-in iterator is actually a reference. This is a bad practice. 🤪